File: apps/relay-server/src/twilio/media_stream.py
"""Twilio Media Stream WebSocket 핸들러.

Twilio Media Stream 이벤트를 수신하고,
수신자 오디오를 Session B로 전달하며,
Session A의 TTS 오디오를 Twilio로 전송한다.
"""

import base64
import json
import logging

from fastapi import WebSocket

from src.types import ActiveCall, TwilioMediaEvent

logger = logging.getLogger(__name__)


class TwilioMediaStreamHandler:
    """Twilio Media Stream WebSocket 연결을 관리한다."""

    def __init__(self, ws: WebSocket, call: ActiveCall):
        self.ws = ws
        self.call = call
        self.stream_sid: str = ""
        self._closed = False

    async def handle_message(self, raw: str) -> TwilioMediaEvent | None:
        """Twilio Media Stream 메시지를 파싱한다."""
        try:
            data = json.loads(raw)
            event = TwilioMediaEvent(**data)
        except Exception:
            logger.warning("Failed to parse Twilio media event: %s", raw[:200])
            return None

        match event.event:
            case "connected":
                logger.info("Twilio media stream connected (call=%s)", self.call.call_id)
            case "start":
                self.stream_sid = event.stream_sid or ""
                self.call.stream_sid = self.stream_sid
                logger.info(
                    "Twilio media stream started: stream_sid=%s", self.stream_sid
                )
            case "media":
                return event  # 오디오 페이로드 — 호출자가 Session B로 전달
            case "stop":
                logger.info("Twilio media stream stopped (call=%s)", self.call.call_id)
                self._closed = True

        return None

    def extract_audio(self, event: TwilioMediaEvent) -> bytes | None:
        """Twilio media 이벤트에서 g711_ulaw 오디오 바이트를 추출한다."""
        if event.media and event.media.get("payload"):
            return base64.b64decode(event.media["payload"])
        return None

    async def send_audio(self, audio_bytes: bytes) -> None:
        """Session A의 TTS 오디오를 Twilio로 전송한다 (g711_ulaw base64)."""
        if self._closed:
            return

        payload = base64.b64encode(audio_bytes).decode("ascii")
        msg = {
            "event": "media",
            "streamSid": self.stream_sid,
            "media": {"payload": payload},
        }
        try:
            await self.ws.send_json(msg)
        except Exception:
            logger.warning("Failed to send audio to Twilio (call=%s)", self.call.call_id)
            self._closed = True

    async def send_clear(self) -> None:
        """Twilio의 오디오 버퍼를 비운다 (interrupt 시 사용)."""
        if self._closed:
            return
        msg = {"event": "clear", "streamSid": self.stream_sid}
        try:
            await self.ws.send_json(msg)
        except Exception:
            pass

    @property
    def is_closed(self) -> bool:
        return self._closed

File: apps/relay-server/src/twilio/media-stream.ts
import type { WebSocket } from 'ws';
import type { TwilioMediaEvent } from '../types.js';

export class TwilioMediaStreamHandler {
  private streamSid: string | null = null;
  private callSid: string | null = null;
  private callId: string | null = null;

  constructor(
    private ws: WebSocket,
    private onAudioReceived: (payload: string, sequenceNumber: number) => void,
    private onStreamStarted: (streamSid: string, callSid: string) => void,
    private onStreamStopped: () => void,
  ) {
    this.ws.on('message', (data) => this.handleMessage(data));
    this.ws.on('close', () => this.onStreamStopped());
    this.ws.on('error', (err) => {
      console.error('[TwilioMediaStream] WebSocket error:', err.message);
    });
  }

  private handleMessage(data: unknown) {
    try {
      const event: TwilioMediaEvent = JSON.parse(
        typeof data === 'string' ? data : (data as Buffer).toString(),
      );

      switch (event.event) {
        case 'start':
          this.streamSid = event.start!.streamSid;
          this.callSid = event.start!.callSid;
          console.log(`[TwilioMediaStream] Stream started: ${this.streamSid}`);
          this.onStreamStarted(this.streamSid!, this.callSid!);
          break;

        case 'media':
          if (event.media?.payload) {
            const seq = parseInt(event.sequenceNumber ?? '0', 10);
            this.onAudioReceived(event.media.payload, seq);
          }
          break;

        case 'stop':
          console.log(`[TwilioMediaStream] Stream stopped: ${this.streamSid}`);
          this.onStreamStopped();
          break;

        case 'mark':
          // Mark events for tracking audio playback completion
          break;
      }
    } catch (err) {
      console.error('[TwilioMediaStream] Failed to parse message:', err);
    }
  }

  sendAudio(payload: string) {
    if (this.ws.readyState !== this.ws.OPEN || !this.streamSid) return;

    this.ws.send(
      JSON.stringify({
        event: 'media',
        streamSid: this.streamSid,
        media: { payload },
      }),
    );
  }

  sendMark(name: string) {
    if (this.ws.readyState !== this.ws.OPEN || !this.streamSid) return;

    this.ws.send(
      JSON.stringify({
        event: 'mark',
        streamSid: this.streamSid,
        mark: { name },
      }),
    );
  }

  clearAudio() {
    if (this.ws.readyState !== this.ws.OPEN || !this.streamSid) return;

    this.ws.send(
      JSON.stringify({
        event: 'clear',
        streamSid: this.streamSid,
      }),
    );
  }

  close() {
    if (this.ws.readyState === this.ws.OPEN) {
      this.ws.close();
    }
  }

  get isConnected(): boolean {
    return this.ws.readyState === this.ws.OPEN;
  }

  get currentStreamSid(): string | null {
    return this.streamSid;
  }
}
